feat: Book & Author

LatifahAlessa opened the pull request:
- add book
- add author
- delete book
- link book to library
- borrow a book
- validate borrowed books
- view all books
- view a specific book -

LatifahAlessa committed:
feat: Book & Author
- add book
- add author
- delete book
- link book to library
- borrow a book
- validate borrowed books
- view all books
- view a specific book
-

HananAlhajri commented on src/main/java/com/example/Library/Management/System/controller/AuthorController.java:
```diff
@@ -0,0 +1,25 @@
+@RequiredArgsConstructor
+@RestController
+@RequestMapping ("/author")
+
```
remove extra space

HananAlhajri commented on src/main/java/com/example/Library/Management/System/controller/BookController.java:
```diff
@@ -0,0 +1,50 @@
+@RequiredArgsConstructor
+@RestController
+@RequestMapping ("/book")
+
```
++

HananAlhajri commented on src/main/java/com/example/Library/Management/System/dto/AuthorDTO.java:
```diff
@@ -0,0 +1,23 @@
+    @Size(max = 20, message = "Last name must be at most 20 characters")
+    private String lastName;
+
+    @PositiveOrZero (message = "Number of books must be 0 or more")
```
Perfect! 👏

HananAlhajri commented on src/main/java/com/example/Library/Management/System/dto/BookSimpleDTO.java:
```diff
@@ -0,0 +1,32 @@
+
+    private Set<Long> authors;
+
+    private Long UserId;
```
follow naming conventions, + clarify variable name for usage, suggestion: `requesterId `

HananAlhajri commented on src/main/java/com/example/Library/Management/System/model/entity/Author.java:
```diff
@@ -0,0 +1,41 @@
+    private Set<Book> books = new HashSet<>();
+
+    @Column
+    private int numberOfBooks;//Do I increment each time i add book to the author?
```
exactly! so this value must not be set when a user adds an author, rather, it's a **transient** value set based on books associated with the author 👌

HananAlhajri commented on src/main/java/com/example/Library/Management/System/model/entity/Book.java:
```diff
@@ -0,0 +1,64 @@
+    private Set<Author> authors = new HashSet<>();
+
+    @Override
+    public boolean equals(Object o) {
```
use Lombok for equals & hash

HananAlhajri commented on src/main/java/com/example/Library/Management/System/model/entity/Library.java:
```diff
@@ -27,7 +29,22 @@ public class Library {
+    private Set<Book> books;
+
+    @Override
+    public boolean equals(Object o) {
```
++

HananAlhajri commented on src/main/java/com/example/Library/Management/System/service/BookService.java:
```diff
@@ -0,0 +1,96 @@
+            bookRepository.delete(book);
+        }
+        else
+            throw new ResponseStatusException(HttpStatus.FORBIDDEN,"Only Admin can delete a book");//is Forbidden a correct exception?
```
Yup! Perfect choice! 💯

HananAlhajri commented on src/main/java/com/example/Library/Management/System/service/BookService.java:
```diff
@@ -0,0 +1,96 @@
+    public void borrowBook(Long id) {
+        Book book = bookRepository.findById(id.intValue()).orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "The book with Id " + id + " is not found"));
+        if (book.isBorrowed())
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "The book is already borrowed");
```
always surround if/else block for better readability

HananAlhajri commented on src/main/java/com/example/Library/Management/System/service/LibraryService.java:
```diff
@@ -57,4 +57,9 @@ public void assignManagerToLibrary (Long libraryId, Long managerId) {//check if
         libraryRepository.save(library);
     }
+
+    public Library getEntityById(Long id) {
```
```suggestion
    public Library getLibraryById(Long id) {
```
or 
```suggestion
    public Library getById(Long id) {
```

HananAlhajri reviewed (changes requested):
Awesome work on this PR, Latifah! 👏

HananAlhajri commented on src/main/java/com/example/Library/Management/System/dto/BookSimpleDTO.java:
```diff
@@ -0,0 +1,32 @@
+
+    private Set<Long> authors;
+
+    private Long requesterId;
```
Since it's a must validate for null

HananAlhajri commented on src/main/java/com/example/Library/Management/System/dto/BookSimpleDTO.java:
```diff
@@ -0,0 +1,32 @@
+
+    private boolean borrowed;
+
+    private Set<Long> authors;
```
Since it's a must validate for empty & null

HananAlhajri commented on src/main/java/com/example/Library/Management/System/controller/BookController.java:
```diff
@@ -0,0 +1,50 @@
+        return ResponseEntity.ok().body("Book is added successfully");
+    }
+
+    @PutMapping("/add-to-library")
```
**PUT** is used when you're updating the whole resource, while **PATCH** is used when you're just updating a part of it.
Think of PUT like replacing the entire file, and PATCH like editing just one section

HananAlhajri commented on src/main/java/com/example/Library/Management/System/controller/BookController.java:
```diff
@@ -0,0 +1,50 @@
+        return ResponseEntity.ok().body("Book is added to the library successfully");
+    }
+
+    @PutMapping("/borrow")
```
same comment above

HananAlhajri commented on src/main/java/com/example/Library/Management/System/service/BookService.java:
```diff
@@ -0,0 +1,107 @@
+        Book book = bookRepository.findById(id.intValue()).orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND,"Book not found"));
+        if (userDTO.getRole() == RoleEnum.ADMIN) {
+            if (book.isBorrowed()) {
+                throw new ResponseStatusException(HttpStatus.FORBIDDEN, "Book is borrowed");
```
`403 Forbidden` is rather used when the user is authenticated/logged in but is not allowed to perform the requested action.

```suggestion
                throw new ResponseStatusException(HttpStatus.UNPROCESSABLE_ENTITY, "Book is borrowed");
```
or
```suggestion
                throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Book is borrowed");
```

HananAlhajri reviewed (changes requested):
I tried to view a specific library yet I didn't get books that I've added 
<img width="972" height="677" alt="image" src="https://github.com/user-attachments/assets/f0465490-ae89-45e8-8a1a-e3c551382864" />

LatifahAlessa committed:
feat: Book & Author
- add book
- add author
- delete book
- link book to library
- borrow a book
- validate borrowed books
- view all books
- view a specific book
-
